TFUtil.py

```diff
@@ -949,8 +962,9 @@ def get_placeholder_flattened(self, keep_dims=False):
       dyn_axes = [0]
     assert dyn_axes == [0]
     if keep_dims and orig_num_dyn_axes >= 2:
-      for i in range(orig_num_dyn_axes - 1):
-        x = tf.expand_dims(x, axis=1)
+      for i in self.get_spatial_batch_axes() + [self.batch_dim_axis]:
+        if i != 0:
+          x = tf.expand_dims(x, axis=i)
```

Comment:
This is confusing.
Better, initially where you have `dyn_axes = self.get_spatial_batch_axes() + [self.batch_dim_axis]`, do instead:
```
orig_dyn_axes = self.get_spatial_batch_axes() + [self.batch_dim_axis]
dyn_axes = list(orig_dyn_axes)
```
And then, here at this place, do this:
```suggestion
for i in orig_dyn_axes:
  if i not in dyn_axes:
    x = tf.expand_dims(x, axis=i)
```
Also, to have another check that we have the correct number of dims, add this after the loop:
```
x.set_shape([None] * self.batch_ndim)
```
Also, please add some test cases for this, for several of the possible cases.